util,sql: add tsvector and tsquery package and sql handling

jordanlewis opened the pull request:
Updates #41288.

This PR adds a subset of Postgres's full text search implementation, which uses the tsvector and tsquery datatypes. It aims to be as compatible as possible with Postgres. The documentation on Postgres's full text search lives here: https://www.postgresql.org/docs/current/textsearch.html

Most of the implementation lives in a new `util/tsearch` package.

Features (all pg-compatible, including the stop words and stemming for the supported languages):

- `tsquery` datatype and query parsing (supports `&`, `|`, `!`, and `<->` (followed by) operators)
- `tsvector` datatype and vector parsing (supports position lists and priorities)
- `to_tsvector` builtin
- `to_tsquery` builtin
- `phraseto_tsquery` builtin
- `plainto_tsquery` builtin
- `tsquery` @@ `tsvector` evaluation 
- the `default_text_search_config` session setting that defaults to `pg_catalog.english`
- stopwords and snowball stemming for english, french, spanish, russian, norwegian, and swedish.
- a "simple" text search configuration that simply normalizes words by lowercasing them
- inverted indexing on tsvector
- inverted index acceleration for `tsvector @@ tsquery` evaluation, which works both on stored computed tsvector columns and computed tsvector indexes

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83736)
<!-- Reviewable:end -->

jordanlewis committed:
util: add tsearch package; tsvector parsing

This commit adds a tsearch package which will contain text search algorithms
for the tsvector/tsquery full text search capability.

It also adds parsing of the tsvector "sub language".

Release note: None

jordanlewis committed:
ts: add parsing for tsquery sublanguage

This commit adds lexing and parsing for the tsquery sublanguage, which
is similar to tsvector but with a few different bells and whistles. The
tsquery is lexed and then parsed into a simple AST.

Release note: None

jordanlewis committed:
tsearch: add eval package

The eval package contains a function to evaluate a tsquery against a
tsvector, in memory.

Release note: None

jordanlewis committed:
tsearch: add encodings for tsvector/tsquery

This commit adds value and pgwire encodings for tsvector and tsquery.

Release note: None

jordanlewis committed:
sql: add tsvector/tsquery types; casts; @@ eval

This commit adds the tsvector and tsquery types to sql, and makes them
available for use using the string::tsvector and string::tsquery casts.
It also implements the in-memory evaluation of @@ (the tsquery match
operator) by delegating to the tsearch.eval package.

Release note (sql change): add in-memory-only evaluation of tsvector and
tsquery datatypes and the @@ matches operator.

jordanlewis committed:
builtins: add to_tsvector {phrase,plain,}to_tsquery

Release note (sql change): add the to_tsvector, to_tsquery,
phraseto_tsquery, and plainto_tsquery builtins which parse input
documents into tsvectors and tsqueries respectively.

jordanlewis committed:
sql: permite inverted idx on tsvector

Release note: None

jordanlewis committed:
sql: add regconfig/regdictionary

Release note: None

jordanlewis committed:
tsearch: add stopwords

jordanlewis committed:
tsearch: add blevesearch snowball stemming

jordanlewis committed:
sql: add default_text_search_config

jordanlewis committed:
sql: inv idx accelerate tsvector@@tsquery queries

jordanlewis commented:
Superseded by a ton of individual PRs.
